File: Kotlin/Solution with CharArray.kt

class Solution {

    private companion object {
        const val ASCII_ZERO = 48
    }

    fun compressedString(word: String): String {
        val compressed = CharArray(2 * word.length)
        var frequency = 1
        var index = 0

        for (i in 1..<word.length) {
            if (frequency == 9 || word[i - 1] != word[i]) {
                compressed[index++] = (frequency + ASCII_ZERO).toChar()
                compressed[index++] = word[i - 1]
                frequency = 1
                continue;
            }
            ++frequency;
        }
        compressed[index++] = (frequency + ASCII_ZERO).toChar()
        compressed[index++] = word[word.length - 1]

        return compressed.slice(0..<index).joinToString("")
    }
}

File: C#/Solution with CharArray.cs

using System;

public class Solution
{
    public string CompressedString(string word)
    {
        var compressed = new char[2 * word.Length];
        int frequency = 1;
        int index = 0;

        for (int i = 1; i < word.Length; ++i)
        {
            if (frequency == 9 || word[i - 1] != word[i])
            {
                compressed[index++] = (char)(frequency + '0');
                compressed[index++] = word[i - 1];
                frequency = 1;
                continue;
            }
            ++frequency;
        }
        compressed[index++] = (char)(frequency + '0');
        compressed[index++] = word[word.Length - 1];

        return String.Join("", compressed.Take(index));
    }
}
